src/main/java/hudson/remoting/AgentFailuresFileListCap.java

```diff
@@ -0,0 +1,83 @@
+package hudson.remoting;
+
+import java.io.File;
+import java.io.FilenameFilter;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Maintains most recent N files in a directory in cooperation with the writer.
+ *
+ * Useful to record incidents as files without bankrupting
+ *
+ */
+public class AgentFailuresFileListCap {
+
+    private static final Logger LOGGER = Logger.getLogger(AgentFailuresFileListCap.class.getName());
+
+    private final File folder;
+    private final LinkedHashSet<File> files = new LinkedHashSet<File>();
+    private int size;
+
+    public AgentFailuresFileListCap(File folder, int size) {
+        this(folder,null,size);
+    }
+
+    public AgentFailuresFileListCap(File folder, FilenameFilter filter, int size) {
+        this.folder = folder;
+        this.size = size;
+
+        if (!folder.exists() && !folder.mkdirs())
+            throw new Error("Failed to create "+folder);
```

Comment:
NIT: just me, but IMO never ever use blocks without braces. Just not worth the risk when touching that code later.